Byron EBB decode accepts nil header; block-fetch of a null body panics the node

gouroboros v0.188.0.

Summary

A block-fetch block message carrying a Byron epoch-boundary block whose body is CBOR null decodes without error into a ByronEpochBoundaryBlock with a nil BlockHeader. Calling any header-backed accessor on that value then dereferences the nil header and panics, so a single malformed or adversarial peer message can crash the receiving node.

Decode path

- protocol/blockfetch/client.go handleBlock (line 519) builds the block via ledger.NewBlockFromCbor (line 575).
- For a Byron EBB block type this reaches ByronEpochBoundaryBlock.UnmarshalCBOR (ledger/byron/byron.go:1201). ByronEpochBoundaryBlock.BlockHeader is a pointer (byron.go:1196). When the encoded body is CBOR null, the decoder treats it as a no-op and leaves BlockHeader nil while UnmarshalCBOR returns no error.
- The accessors then panic: ByronEpochBoundaryBlock.SlotNumber (byron.go:1232) calls ByronEpochBoundaryBlockHeader.SlotNumber (byron.go:1091), which reads h.ConsensusData.Epoch. With a nil header this is a nil dereference.

Observed panic

panic: runtime error: invalid memory address or nil pointer dereference [signal SIGSEGV code=0x1 addr=0x58]
  ledger/byron.(*ByronEpochBoundaryBlockHeader).SlotNumber   byron/byron.go:1092
  ledger/byron.(*ByronEpochBoundaryBlock).SlotNumber         byron/byron.go:1233
  protocol/blockfetch.(*Client).handleBlock                  blockfetch/client.go:601

addr=0x58 matches the offset of ConsensusData.Epoch within the header.

Impact

Any peer over the block-fetch mini-protocol can crash a node by sending a block message with a null or otherwise header-less Byron EBB body. This is a liveness/availability problem, not silent divergence.

Suggested fix

Reject the malformed input during decode instead of producing a partially-nil value:
- In ByronEpochBoundaryBlock.UnmarshalCBOR (and the Byron main block equivalent), return an error when the decoded header pointer is nil.
- Optionally nil-guard the Byron EBB accessors (SlotNumber, BlockNumber, Hash, IssuerVkey, PrevHash) so a header-less block cannot panic a caller.

ByronMainBlock also holds BlockHeader as a pointer (byron.go:1125) and should get the same nil-header check.

--- a/ledger/byron/block_test.go
+++ b/ledger/byron/block_test.go
@@ -90,6 +90,19 @@ func TestByronBlock_CborRoundTrip_UsingCborEncode(t *testing.T) {
 	}
 }
 
+func TestByronEpochBoundaryBlockNullHeader(t *testing.T) {
+	// [null, [], []]: an EBB block whose header is CBOR null
+	dataBytes, err := hex.DecodeString("83f68080")
+	if err != nil {
+		t.Fatalf("Failed to decode hex string into CBOR bytes: %v", err)
+	}
+	var block byron.ByronEpochBoundaryBlock
+	err = block.UnmarshalCBOR(dataBytes)
+	if err == nil {
+		t.Fatal("expected error decoding EBB block with null header, got none")
+	}
+}
+
 func TestByronTransaction_Utxorpc(t *testing.T) {
 	// Decode the hex string to bytes
 	blockBytes, err := hex.DecodeString(strings.TrimSpace(testdata.ByronBlockHex))

--- a/ledger/byron/byron.go
+++ b/ledger/byron/byron.go
@@ -1204,6 +1204,9 @@ func (b *ByronEpochBoundaryBlock) UnmarshalCBOR(cborData []byte) error {
 	if _, err := cbor.Decode(cborData, &tmp); err != nil {
 		return err
 	}
+	if tmp.BlockHeader == nil {
+		return errors.New("byron EBB block missing header")
+	}
 	*b = ByronEpochBoundaryBlock(tmp)
 	b.SetCbor(cborData)
 	return nil

